kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/algebras/DoubleTensorAlgebra.kt

```diff
@@ -0,0 +1,569 @@
+/*
+ * Copyright 2018-2021 KMath contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package space.kscience.kmath.tensors.core.algebras
+
+import space.kscience.kmath.nd.as2D
+import space.kscience.kmath.tensors.api.TensorPartialDivisionAlgebra
+import space.kscience.kmath.tensors.api.Tensor
+import space.kscience.kmath.tensors.core.*
+import space.kscience.kmath.tensors.core.algebras.DoubleAnalyticTensorAlgebra.fold
+import space.kscience.kmath.tensors.core.algebras.DoubleAnalyticTensorAlgebra.foldDim
+import space.kscience.kmath.tensors.core.broadcastOuterTensors
+import space.kscience.kmath.tensors.core.checkBufferShapeConsistency
+import space.kscience.kmath.tensors.core.checkEmptyDoubleBuffer
+import space.kscience.kmath.tensors.core.checkEmptyShape
+import space.kscience.kmath.tensors.core.checkShapesCompatible
+import space.kscience.kmath.tensors.core.checkTranspose
+import space.kscience.kmath.tensors.core.checkView
+import space.kscience.kmath.tensors.core.dotHelper
+import space.kscience.kmath.tensors.core.getRandomNormals
+import space.kscience.kmath.tensors.core.minusIndexFrom
+import kotlin.math.abs
+
+/**
+ * Implementation of basic operations over double tensors and basic algebra operations on them.
+ */
+public open class DoubleTensorAlgebra : TensorPartialDivisionAlgebra<Double> {
```

Comment:
This solution is bad since we want stateless objects to be closed, It enables additional optimizations from the JVM. It this class will never be used, better to make it abstract.

Reply:
you should use `BroadcastDoubleTensorAlgebra` in that case